airbyte-workers/src/main/java/io/airbyte/workers/temporal/PartialSuccessTemporalAttemptExecution.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.workers.temporal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.functional.CheckedConsumer;
+import io.airbyte.commons.functional.CheckedSupplier;
+import io.airbyte.scheduler.models.JobRunConfig;
+import io.airbyte.workers.Worker;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/*
+ * This class allow a worker to run multiple times. In addition to the functionality in
+ * TemporalAttemptExecution it takes a predicate to determine if the output of a worker constitutes
+ * a complete success or a partial one. It also takes a function that takes in the input of the
+ * previous run of the worker and the output of the last worker in order to generate a new input for
+ * that worker.
+ */
+public class PartialSuccessTemporalAttemptExecution<INPUT, OUTPUT> implements Supplier<List<OUTPUT>> {
```

Comment:
would a better name for this be `AutoRetryingTemporalAttemptExecution` or something? I think as we make this part of the protocol, the fact that we retain state after failures should hopefully be self evident